### PR TITLE
Multiple new Schwab transaction types

### PR DESCRIPTION
**Pull request by Zburatorul**
This adds support for five new transaction types.

The way I dealt with `STOCKMERGER` by adding new fields to `BankingEntryType` and modifying the loop in `_load_brokerage_transaction` is ugly. Better suggestions welcome.
Furthermore, it is unclear what cost basis to use for the lot that's being reduced so I insert "FIXME" instead which at least gets the transaction in the journal and prompts the user to choose a lot manually.

### CODE REVIEW

File: beancount_import/source/schwab_csv.py
Comment (bayesianmind) [inline beancount_import/source/schwab_csv.py:1308]: If you wanted to avoid this one idea is to modify the loop like I do here: https://github.com/bayesianmind/beancount-import/blob/master/beancount_import/source/schwab_equityaward.py#L108

I think the way you've done it here is the least bad option FWIW.

File: beancount_import/source/schwab_csv.py
Comment (bayesianmind) [inline beancount_import/source/schwab_csv.py:229]: This isn't ideal as you note.
It's beyond the scope of this PR, but given Schwab's extremely limited CSV format requiring them to do ugly things like break a single transaction into multiple rows it is better to have flexibility when parsing. If we move away from the thin simple parser that just dumps CSV rows into `RawBrokerageEntry` and instead move the logic of `get_processed_entry` into the CSV reading logic it would be a net improvement (in this programmer's opinion). If we had that structure you'd construct the Merger data class in the CSV parsing loop.

The approach you take here isn't terrible though and I would be fine with it if it were my codebase. If we ended up finding several more transaction types like this then it will get really ugly, but I think it's unlikely there are too many transaction types that span multiple entries like this.

**Review by bayesianmind — APPROVED** [review]

File: beancount_import/source/schwab_csv.py
Comment (Zburatorul) [reply beancount_import/source/schwab_csv.py:229]: I think we should defer the act moving the logic you mention into the CSV reading logic until another such situation arises. It will at least be more clear how to do it then.

Comment (bayesianmind) [reply beancount_import/source/schwab_csv.py:229]: Makes sense to me.

Comment (carljm) [reply beancount_import/source/schwab_csv.py:229]: I'm not sure it really helps to move more logic into CSV parsing. Either way we end up having to do the same thing to handle a case like this (loop over sequential rows with some aggregation logic). The intention of the design is that the `Raw*` objects should be a very close mirror of exactly what is in the CSV file, and then in a separate step we process those raw entries into something more useful for importing into beancount. This separates the CSV-parsing details from the logical interpretation of the CSV rows actually mean. To me this seems like useful separation of concerns, and I don't see what is gained by collapsing them into a single step.

File: beancount_import/source/schwab_csv.py
Comment (carljm) [inline beancount_import/source/schwab_csv.py:342]: rather than duplicating the creation of the shared_attrs dict, let's follow the existing pattern of `STOCK_PLAN_ACTIVITY` which also needs to do some prior munging of `amount`, and have two `if self.action is BrokerageAction.EXPIRED`, one before the construction of `shared_attrs` and one after.

File: beancount_import/source/schwab_csv.py
Comment (carljm) [inline beancount_import/source/schwab_csv.py:354]: again here it looks like we are re-handling things that are already handled in `shared_attrs`, and then not using `shared_attrs`. This makes future maintenance more difficulty; if we need to make global adjustments that apply to a certain shared attribute of all transaction types, it will appear that we can make this change in `shared_attrs`, and we may not notice that some transaction types aren't using it. So I'd rather use the existing pattern of `STOCK_PLAN_ACTIVITY` as mentioned above to ensure that all transaction types actually do share `shared_attrs`. If we have to use that pattern too many times for some field, we can move that field out of `shared_attrs` and instead pass it individually for each transaction type.

File: beancount_import/source/schwab_csv.py
Comment (carljm) [inline beancount_import/source/schwab_csv.py:628]: We could introduce a `get_other_meta()` method (defaulting to `return {}` but overridden here to `return self.get_meta()`) so that you don't have to override all of `get_postings()` and repeat a lot of what is already done in the base method. It looks like all the other differences here can already be handled by existing methods e.g. `get_other_units()` and `get_cost()`.

I don't feel strongly about this though, this override is OK.

File: beancount_import/source/schwab_csv.py
Comment (carljm) [reply beancount_import/source/schwab_csv.py:1308]: I agree, this doesn't seem bad at all to me. I think it makes sense in this case to do this line merging here in the CSV parsing, even though it's a slight departure from the "RawEntries are 1:1 representations of rows in the CSV" design.

One small tweak I might make is to store the three items as a tuple or NamedTuple, so that you can just have a single variable to set to either None or that tuple, and then not have to arbitrarily choose one of them to check for None-ness.

**Review by carljm — COMMENTED** [review]
Looks good, thanks! A few small suggestions, but the handling of mergers seems fine overall.

File: beancount_import/source/schwab_csv.py
Comment (Zburatorul) [reply beancount_import/source/schwab_csv.py]: Introduced the nametuple.

File: beancount_import/source/schwab_csv.py
Comment (Zburatorul) [reply beancount_import/source/schwab_csv.py]: Done.

File: beancount_import/source/schwab_csv.py
Comment (Zburatorul) [reply beancount_import/source/schwab_csv.py]: Done.

File: beancount_import/source/schwab_csv.py
Comment (Zburatorul) [reply beancount_import/source/schwab_csv.py]: Eh, I prefer less code so I removed the override.

**Comment by Zburatorul** [comment]
I have made the changes suggested by the latest comments and this is ready to merge, unless more objections.

**Review by carljm — APPROVED** [review]
Looks good to me! Fine with me to merge it now and update with lot info later, or wait to merge and update it, either way.

**Comment by Zburatorul** [comment]
Rebased and got tox to pass. Ready to merge.
Will get it to work with latest lot capabilities in separate PR.

**Review by carljm — CHANGES_REQUESTED** [review]
A few review comments, mostly nits that I'm fine with either way, but the duplicate `EXPIRED` blocks and the changes in `test_lots/` should be fixed.

File: testdata/source/schwab_csv/test_lots/positions/Intelligent-Positions-2020-10-25.CSV
Comment (Zburatorul) [reply testdata/source/schwab_csv/test_lots/positions/Intelligent-Positions-2020-10-25.CSV]: Indeed. Fixed.

File: beancount_import/source/schwab_csv.py
Comment (Zburatorul) [reply beancount_import/source/schwab_csv.py]: Done.

File: beancount_import/source/schwab_csv.py
Comment (Zburatorul) [reply beancount_import/source/schwab_csv.py]: Rebase conflict artifact.

File: beancount_import/source/schwab_csv.py
Comment (Zburatorul) [reply beancount_import/source/schwab_csv.py]: Changed for consistency.

File: beancount_import/source/schwab_csv.py
Comment (Zburatorul) [reply beancount_import/source/schwab_csv.py:183]: Let's do that.

File: beancount_import/source/schwab_csv.py
Comment (Zburatorul) [reply beancount_import/source/schwab_csv.py]: Changed it to a dataclass. One odd thing is that quantity must be Optional[Decimal] for mypy to pass (because that's what _convert_decimal returns), but in reality it really cannot be None. This means in such situations we end up inserting asserts right before the creation of the dataclass. I don't know of a better pattern. Let me know if you do.

File: beancount_import/source/schwab_csv.py
Comment (carljm) [inline beancount_import/source/schwab_csv.py:385]: This `lot_info` variable is not used.

File: beancount_import/source/schwab_csv.py
Comment (carljm) [inline beancount_import/source/schwab_csv.py:381]: very minor nit, but the rest of the cases in this method use `==` not `is`. I think both should work equally well in this case but we may as well be consistent

File: beancount_import/source/schwab_csv.py
Comment (carljm) [inline beancount_import/source/schwab_csv.py:383]: I think the wrong version of this block was kept after merge conflict, and we've now lost some of my earlier feedback. Notably we are redefining `amount` here after already having defined it above, and we are not using `**shared_attrs` below.

File: beancount_import/source/schwab_csv.py
Comment (carljm) [inline beancount_import/source/schwab_csv.py:887]: I think this could be resolvable via `lot_info`, but it's OK to merge without implementing that yet.

**Review by carljm — CHANGES_REQUESTED** [review]
Sorry, saw a few things yet in this updated version.

File: beancount_import/source/schwab_csv.py
Comment (Zburatorul) [reply beancount_import/source/schwab_csv.py]: Good catch.

File: beancount_import/source/schwab_csv.py
Comment (Zburatorul) [reply beancount_import/source/schwab_csv.py:916]: I want to handle this in a separate PR. This PR is big as is.

File: beancount_import/source/schwab_csv.py
Comment (Zburatorul) [reply beancount_import/source/schwab_csv.py]: Changed

File: beancount_import/source/schwab_csv.py
Comment (Zburatorul) [reply beancount_import/source/schwab_csv.py]: Passed it to Sell().

Comment (Zburatorul) [reply beancount_import/source/schwab_csv.py]: Anything to do with lots I'm leaving for future PRs.

File: testdata/source/schwab_csv/test_lots/positions/Intelligent-Positions-2020-10-25.CSV
Comment (carljm) [reply testdata/source/schwab_csv/test_lots/positions/Intelligent-Positions-2020-10-25.CSV:4]: It seems like these changes in test_lots are still here? I made this 10.5 intentionally to validate that the lot details handling allows for fractional shares.

**Review by carljm — APPROVED** [review]
Looks good, thank you!!

**Pull request merged**
